File: convertedToIOS5/p741p758XML/p741p758XML/MyXMLParserDelegate.m


#import "MyXMLParserDelegate.h"


@implementation MyXMLParserDelegate
@synthesize text, parent, child, name;

- (void) start: (NSString*) el parent: (id) p {
    self.name = el;
    self.parent = p;
    self.text = [NSMutableString string];
}

- (void) makeChild: (Class) class elementName: (NSString*) elementName parser: (NSXMLParser*) parser {
    MyXMLParserDelegate* del = [[class alloc] init];
    self.child = del;
    parser.delegate = del;
    [del start: elementName parent: self];
}

- (void) finishedChild: (NSString*) s { // subclass implements as desired
}

- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string {
    [self.text appendString:string];
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName {
    if (parent) {
        [parent finishedChild: [self.text copy]];
        parser.delegate = self.parent;
    }
}



@end

File: iOS9bookExamples/bk2ch23p818xml/ch36p1070xml/MyXMLParserDelegate.swift


import UIKit


class MyXMLParserDelegate : NSObject {
    var name : String!
    var text = ""
    weak var parent : MyXMLParserDelegate?
    var child : MyXMLParserDelegate!
    
    required init(name:String, parent:MyXMLParserDelegate?) {
        self.name = name
        self.parent = parent
        super.init()
    }
    
    func makeChild(klass:MyXMLParserDelegate.Type, elementName:String, parser:NSXMLParser) {
        let del = klass.init(name:elementName, parent:self)
        self.child = del
        parser.delegate = del
    }
    func finishedChild(s:String) {
        fatalError("Subclass must implement finishedChild:!")
    }
    
}

extension MyXMLParserDelegate : NSXMLParserDelegate {
    func parser(parser: NSXMLParser, foundCharacters string: String) {
        self.text = self.text + string
    }
    
    func parser(parser: NSXMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if self.parent != nil {
            self.parent!.finishedChild(self.text)
            parser.delegate = self.parent
        }
    }
    
}


